keps/sig-cloud-provider/2699-add-webhook-hosting-to-ccm/README.md

```diff
@@ -0,0 +1,455 @@
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP will detail enhancing the CCM framework to support cloud provider
+specific webhooks. The intent is to make it easy to either generate a binary
+or enhance the existing CCM binary to host such webhooks. We also intend to
+allow for easily linking in "standard" webhooks needed by other SIGs which
+need to be customized for particular cloud providers.
+
+## Motivation
+
+The Cloud (Provider) Controller Manager (CCM) is the binary into which the Cloud
+Provider places all the controllers needed to make a Kubernetes cluster work
+correctly on their Cloud. There are also occasions when it makes sense for a Cloud
+Provider to want these customizations to be applied inline rather than 
+asynchronously after a change has already been applied. 
+
+Our initial example of this is from SIG Storage. These would like the functionality 
+from PVL admission controller (https://github.com/kubernetes/kubernetes/issues/52617).
+This needs to be completed for cloud provider extraction to complete. Several
+Cloud Providers have indicated that this should be done inline, especially as the
+existing deprecated solution is an inline solution.
+
+### Goals
+
+Our current in tree Cloud Providers are able to stop using the existing PVL admission 
+controller and do so using the framework.
+
+### Non-Goals
+
+We are not intending to create a general admission webhook solution. This is just
+intended to host Cloud Provider specific webhooks as part of the Control Plane.
+
+## Proposal
+
+We will start by adding extension hooks which can be registered in the 
+cmd/cloud-controller-manager/main.go. This would be similar to the mechanism
+we already use to register new controllers. The existing sample shows this with
+a sample of registering the nodeipamcontroller which is not a normally installed
+controller in the cloud controller manager. In a similar way we will have a 
+sample of integrating a PVL mutating webhook into the sample CCM. We will also 
+have the system automatically detect if there are both controllers and webhooks
+registered in the binary. If both are registered it will automatically add 
```

Comment:
Will registration of the MutatingWebhookConfiguration or ValidatingWebhookConfiguration be part of this or is that up to the cluster admin to configure?

Reply:
Great point. I think we should add code to automatically register the webhook as a (optional/beta) part of the webhook server coming up. Will add that to the KEP.

Reply:
However I believe this registration has to be optional, as an escape hatch for things like non standard call paths or the need to add things like labels/label-selectors.

Reply:
Added